cwlVersion: v1.0
class: Workflow

requirements:
  - class: SubworkflowFeatureRequirement
  - class: ScatterFeatureRequirement
  - class: StepInputExpressionRequirement
  - class: InlineJavascriptRequirement
  - class: MultipleInputFeatureRequirement


inputs:

  fastq_file_1:
    type: File
    doc: "Paired-end sequencing data 1 in FASTQ format (fastq, fq, bzip2, gzip, zip)"

  fastq_file_2:
    type: File
    doc: "Paired-end sequencing data 2 in FASTQ format (fastq, fq, bzip2, gzip, zip)"

  indices_folder:
    type: Directory
    doc: "Directory with the genome indices generated by Bowtie"

  blacklisted_regions_bed:
    type: File
    doc: "Blacklisted genomic regions file in BED format"

  genome_fasta_file:
    type: File
    secondaryFiles:
      - .fai
    doc: "Reference genome sequence FASTA and FAI index files"

  genome_size:
    type: string
    doc: "The length of the mappable genome (hs, mm, ce, dm or number, for example 2.7e9)"

  threads:
    type: int?
    default: 4
    doc: "Number of threads for those steps that support multithreading"


outputs:

  bambai_pair:
    type: File
    outputSource: sort_and_index_bam_after_filtering/bam_bai_pair
  
  merge_peaks_with_counts:
    type: File
    outputSource: count_tags/intersected_file

  merge_peaks_sequences:
    type: File
    outputSource: get_sequences/sequences_file


steps:

  qc_fastq_1:
    run: ../tools/fastqc.cwl
    in:
      reads_file: fastq_file_1
    out:
      - summary_file
      - html_file

  qc_fastq_2:
    run: ../tools/fastqc.cwl
    in:
      reads_file: fastq_file_1
    out:
      - summary_file
      - html_file

  trim_read_adapters:
    run: ../tools/trimgalore.cwl
    in:
      input_file: fastq_file_1
      input_file_pair: fastq_file_2
      dont_gzip:
        default: true
      length:
        default: 30
      trim1:
        default: true
      paired:
        default: true
    out:
      - trimmed_file
      - trimmed_file_pair
      - report_file
      - report_file_pair

  rename_fastq_1:
    run: ../tools/rename.cwl
    in:
      source_file: trim_read_adapters/trimmed_file
      target_filename:
        source: fastq_file_1
        valueFrom: $(self.basename)
    out: [target_file]

  rename_fastq_2:
    run: ../tools/rename.cwl
    in:
      source_file: trim_read_adapters/trimmed_file_pair
      target_filename:
        source: fastq_file_2
        valueFrom: $(self.basename)
    out: [target_file]

  align_reads:
    run: ../tools/bowtie2.cwl
    in:
      filelist: rename_fastq_1/target_file
      filelist_mates: rename_fastq_2/target_file
      indices_folder: indices_folder
      end_to_end_very_sensitive:
        default: true
      maxins:
        default: 2000
      no_discordant:
        default: true
      no_mixed:
        default: true
      threads: threads
    out:
      - output
      - output_log

  sort_and_index_bam:
    run: ../tools/samtools-sort-index.cwl
    in:
      sort_input: align_reads/output
      threads: threads
    out: [bam_bai_pair]

  remove_chromosome:
    run: ../tools/samtools-filter.cwl
    in:
      bam_bai_pair: sort_and_index_bam/bam_bai_pair
      exclude_chromosome:
        default: "chrM chrY chrX"
    out: [filtered_bam_bai_pair]
  
  preseq:
    run: ../tools/preseq-lc-extrap.cwl
    in:
      bam_file: remove_chromosome/filtered_bam_bai_pair
      pe_mode:
        default: true
      extrapolation:
        default: 1000000000
    out: [estimates_file]

  alignment_statistics:
    run: ../tools/samtools-stats.cwl
    in:
      bambai_pair: remove_chromosome/filtered_bam_bai_pair
    out:
      - log_file
      - average_length    

  filter_bam:
    run: ../tools/deeptools-alignmentsieve.cwl
    in:
      bambai_pair: remove_chromosome/filtered_bam_bai_pair
      blacklisted_regions: blacklisted_regions_bed
      min_mapping_quality: 
        default: 30
      atac_shift:
        default: true
      ignore_duplicates:
        default: true
      threads: threads
    out: [filtered_bam_file]

  sort_and_index_bam_after_filtering:
    run: ../tools/samtools-sort-index.cwl
    in:
      sort_input: filter_bam/filtered_bam_file
      threads: threads
    out: [bam_bai_pair]

  call_peaks:
    run: ../tools/macs2-callpeak.cwl
    in:
      treatment_file: sort_and_index_bam_after_filtering/bam_bai_pair
      genome_size: genome_size
      shift:
        source: alignment_statistics/average_length
        valueFrom: $(-Math.round(self/2))
      nomodel:
        default: true
      extsize: alignment_statistics/average_length
      format_mode:
        default: "BAM"
      keep_dup:
        default: "all"
    out: [narrow_peak_file]

  sort_peaks:
    run: ../tools/linux-sort.cwl
    in:
      unsorted_file: call_peaks/narrow_peak_file
      key:
        default: ["1,1","2,2n"]
    out: [sorted_file]

  merge_peaks:
    run: ../tools/bedtools-merge.cwl
    in:
      bed_file: sort_peaks/sorted_file
    out: [merged_bed_file]

  count_tags:
    run: ../tools/bedtools-intersect.cwl
    in:
      file_a: merge_peaks/merged_bed_file
      file_b: sort_and_index_bam_after_filtering/bam_bai_pair
      count:
        default: true
    out: [intersected_file]

  get_sequences:
    run: ../tools/bedtools-getfasta.cwl
    in:
      genome_fasta_file: genome_fasta_file
      intervals_file: merge_peaks/merged_bed_file
    out: [sequences_file]